File: 368-largest-divisible-subset/largest-divisible-subset.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        int n=nums.size();

        vector<int> lis(n,1); //lis length
        vector<int> preIndex(n,-1); //lis preIndex

        sort(nums.begin(),nums.end());
     
        for(int i=0;i<n;i++){
            for(int j=i-1;j>=0;j--){
                if((nums[i]%nums[j])==0 && (lis[i]<lis[j]+1) ){
                    lis[i]=lis[j]+1;
                    preIndex[i]=j;
                }
            }
        }
        int maxlis=0;
        int ind=0;
        for(int i=0;i<n;i++){
            if(maxlis<lis[i]){
                maxlis=lis[i];
                ind=i;
            }
        }
        
        vector<int>res;
        while(ind !=-1){
            res.push_back(nums[ind]);
            ind=preIndex[ind];
        }
        reverse(res.begin(),res.end());
        return res;
    }
};

File: 368-largest-divisible-subset/largest-divisible-subset.py
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
     nums.sort()
     n=len(nums)
     maxlcsi=0
     maxlcs=1
     temp=[i for i in range(n)]
     dp=[1 for i in range(n)]
     for i in range(1,n):
      for prev in range(i):
       if nums[i]%nums[prev]==0 and dp[i]<dp[prev]+1:
         dp[i]=dp[prev]+1
         temp[i]=prev
      if maxlcs<dp[i]:
       maxlcsi=i
       maxlcs=dp[i]
     li=[nums[maxlcsi]]
     while temp[maxlcsi] !=maxlcsi:
      maxlcsi=temp[maxlcsi]
      li.append(nums[maxlcsi])
     li.reverse() 
     return li